js/symbol/collision_tile.js

```diff
@@ -52,61 +66,95 @@ CollisionTile.prototype.placeCollisionFeature = function(collisionFeature) {
-            var s1 = (blocking.x1 - box.x2) / (x - blockingAnchorPoint.x); // scale at which new box is to the left of old box
-            var s2 = (blocking.x2 - box.x1) / (x - blockingAnchorPoint.x); // scale at which new box is to the right of old box
-            var s3 = (blocking.y1 - box.y2) * yStretch / (y - blockingAnchorPoint.y); // scale at which new box is to the top of old box
-            var s4 = (blocking.y2 - box.y1) * yStretch / (y - blockingAnchorPoint.y); // scale at which new box is to the bottom of old box
+                minPlacementScale = this.checkForCollision(minPlacementScale, anchorPoint, box, blockingAnchorPoint, blocking);
+                if (minPlacementScale >= this.maxScale) {
+                    return minPlacementScale;
+                }
+            }
+        }
 
-            if (isNaN(s1) || isNaN(s2)) s1 = s2 = 1;
-            if (isNaN(s3) || isNaN(s4)) s3 = s4 = 1;
+        if (avoidEdges) {
+            var reverseRotationMatrix = this.reverseRotationMatrix;
+            var tl = new Point(box.x1, box.y1).matMult(reverseRotationMatrix);
+            var tr = new Point(box.x2, box.y1).matMult(reverseRotationMatrix);
+            var bl = new Point(box.x1, box.y2).matMult(reverseRotationMatrix);
+            var br = new Point(box.x2, box.y2).matMult(reverseRotationMatrix);
+            var rotatedCollisionBox = new CollisionBox(box.anchorPoint,
+                    Math.min(tl.x, tr.x, bl.x, br.x),
+                    Math.min(tl.y, tr.x, bl.x, br.x),
+                    Math.max(tl.x, tr.x, bl.x, br.x),
+                    Math.max(tl.y, tr.x, bl.x, br.x),
+                    box.maxScale);
+
+            for (var k = 0; k < this.edges.length; k++) {
+                var edgeBox = this.edges[k];
+                minPlacementScale = this.checkForCollision(minPlacementScale, box.anchorPoint, rotatedCollisionBox, edgeBox.anchorPoint, edgeBox);
+                if (minPlacementScale >= this.maxScale) {
+                    return minPlacementScale;
+                }
+            }
+        }
+    }
 
-            var collisionFreeScale = Math.min(Math.max(s1, s2), Math.max(s3, s4));
+    return minPlacementScale;
+};
 
-            if (collisionFreeScale > blocking.maxScale) {
-                // After a box's maxScale the label has shrunk enough that the box is no longer needed to cover it,
-                // so unblock the new box at the scale that the old box disappears.
-                collisionFreeScale = blocking.maxScale;
-            }
 
-            if (collisionFreeScale > box.maxScale) {
-                // If the box can only be shown after it is visible, then the box can never be shown.
-                // But the label can be shown after this box is not visible.
-                collisionFreeScale = box.maxScale;
-            }
+CollisionTile.prototype.checkForCollision = function(minPlacementScale, anchorPoint, box, blockingAnchorPoint, blocking) {
```

Comment:
I suppose keeping these method names as similar as possible to those in native is most important.